### DISCUSSION TITLE
Remove `defer` attribute from README

### DISCUSSION THREAD

**Discussion by ryangjchandler** [alpinejs/alpine#884] [3. Feature ideas]
Opening up this discussion to get everyone's thoughts on the above.

Currently, the README suggests placing the script in the `<head>` of the document with a `defer` attribute. This is fine when you're just running Alpine on it's own, but when you start introducing Livewire or third party scripts like Spruce or Magic Helpers, it becomes problematic and causes problems since those third parties require Alpine to be loaded _after_ their own scripts.

I'm suggesting that we instead update the documentation and instruct users to place Alpine at the end of their `<body>`, without the `defer` attribute, that way those third-parties don't need to instruct people about removing the README recommendation. There's also **a lot** of people that come into the Discord with similar problems and the answer is to remove the `defer` attribute and either have render-blocking `<script>` tags in the `<head>` or move it to the end of the body.

**Comment by KevinBatdorf** [comment]
Asking users to remove `defer` does seem to be one of those solutions that works when nothing else makes sense. From what I can tell, this is related to the browser not always being consistent with timing on rendering deferred assets, even though the spec is clear on it. 

The only issue with removing `defer` would be that scripts with the type set to `module` are deferred as well, so there is some consistency lost. That may not matter, but it might be better to have a warning note on the install section instead.

I wonder if there is an opportunity to improve Alpine here though. For example, if a required function is undefined, check again a bit later.

**Comment by pomartel** [comment]
Be careful that putting the script in the body might not be compatible with [Turbolinks](https://github.com/turbolinks/turbolinks#loading-your-applications-javascript-bundle). I know this is still a very popular library in the Rails world (I use Turbolinks in all my project) and makes for a great combination with Alpine as opposed to building a SPA.

I use deferred script with Alpine in my app and it works great. But since I load Alpine before my application scripts, I had to defer starting Alpine using the `deferLoadingAlpine` function. This method is not explained anywhere in the README (I had to dig in the source code to find it!).

**Reply by ryangjchandler** [reply]
@pomartel Yeah, the `deferLoadingAlpine` hook is something that isn't documented but also used by Spruce and the Magic Helpers package, as well as Livewire.

To be fair, I fixed Spruce so that it now works with Alpine in the `<head>` and `defer`.

**Reply by KevinBatdorf** [reply]
The readme might be a good place to explain all of this, so each user can determine the best loading strategy and Alpine remains neutral. 

Or even this discussion might be a good reference to those who are searching Google for answers? Maybe share some snippets here?

**Reply by ryangjchandler** [reply]
So the main problem with Spruce was the fact that I wasn't setting `window.Spruce = Spruce` until the `deferLoadingAlpine` block ran. I was also `await domReady()` in the `Spruce.start()` function which was causing some problems since Alpine also runs the same bit of `async` code.

Without those bits of logic, it all works fine.

**Reply by pomartel** [reply]
I just don't know if, in general, recommending to put scripts at the end of the body is still considered a best practice for web development. I used to do it 10 years ago but since then I think all major browsers have good support for defer and will respect the defer load order.

I haven't been part of the Alpine community for long  but what I saw is people using `async` in their scripts or loading the Alpine script before their applications scripts (like I did!). I'm not convinced that the root cause of the problem is with defer.

**Reply by pomartel** [reply]
This is a simplified example of what worked for me:

````html
<head>
<script>
window.deferLoadingAlpine = function(callback) {
    window.addEventListener("load", function(event) {
        callback()
    }
}
</script>
<script src="https://cdn.jsdelivr.net/gh/alpinejs/alpine@v2.x.x/dist/alpine.min.js" defer></script>
<script src="/my_app_code.js" defer></script>
</head>
<body>
    <div x-data='myComponent()'></div>
</body>
````
Without using `deferLoadingAlpine`, Alpine would try to call `myComponent()` before the function would be loaded in `my_app_code.js`. Now I think reversing the load order of the two scripts would also work but this would have been too big of a change in the way my app was packaged.

**Reply by ryangjchandler** [reply]
Correct, surely the fix there would be putting `my_app_code.js` before Alpine then?

**Reply by KevinBatdorf** [reply]
The problem with this approach is that you're telling Alpine to wait until the `load` event instead of `DOMContentLoaded`, which could be very late on sites with a lot of images.

> Correct, surely the fix there would be putting my_app_code.js before Alpine then?

This should work 999/1000 times but the 1 time it doesn't, and they show up in Discord with `functionName is undefined` the remedy is to remove defer, and it works.

(oddly enough, even with the errors in the console, Alpine typically will still work)

**Reply by pomartel** [reply]
Yes, so reality is always a bit messier than the examples! 😁 My bundles looks more like this:
````html
<script src="/vendor.js" defer></script>
<script src="/application.js" defer></script>
````
All third party libs are bundled in `vendor.js` and so I couldn't change the load order just for Alpine.

**Comment by ekwoka** [comment]
This is old, but `defer` is the better choice.

Then those others can be run without defer or with `defer` before Alpine.

`defer` and `type=module` run in the order they're defined in all browsers I've been able to check, at least now on 2024.
